src/main/scala/viper/gobra/translator/encodings/slices/SliceEncoding.scala

```diff
@@ -0,0 +1,461 @@
+
+        ctx.option.some(maxOptT match {
+          case None => sliceFromSlice(vpr.Ref, baseOptT, lowT, highT)(ctx)(pos, info, errT)
+          case Some(maxT) => fullSliceFromSlice(vpr.Ref, baseOptT, lowT, highT, maxT)(ctx)(pos, info, errT)
+        })(pos, info, errT)
+      }
+
+      case (lit : in.SliceLit) :: ctx.Slice(_) => {
+        val litA = lit.asArrayLit
+        val tmp = in.LocalVar(Names.freshName, litA.typ.withAddressability(Addressability.pointerBase))(lit.info)
+        val tmpT = ctx.typeEncoding.variable(ctx)(tmp)
+        for {
+          initT <- ctx.typeEncoding.initialization(ctx)(tmp)
+          assignT <- ctx.typeEncoding.assignment(ctx)(in.Assignee.Var(tmp), litA, lit)
+          sliceT <- ctx.expr.translate(in.Slice(tmp, in.IntLit(0)(lit.info), in.IntLit(litA.length)(lit.info), None)(lit.info))(ctx)
+          _ <- local(tmpT)
+          _ <- write(initT)
+          _ <- write(assignT)
+        } yield sliceT
+      }
+    }
+  }
+
+  override def reference(ctx : Context) : in.Location ==> CodeWriter[vpr.Exp] = default(super.reference(ctx)) {
+    case (exp @ in.IndexedExp(base :: ctx.Slice(typ), idx)) :: _ / Shared => for {
+      baseT <- ctx.expr.translate(base)(ctx)
+      idxT <- ctx.expr.translate(idx)(ctx)
+      typT = ctx.typeEncoding.typ(ctx)(typ)
+    } yield {
+      val (pos, info, errT) = exp.vprMeta
+      val getT = ctx.option.get(baseT, ctx.slice.typ(typT))(pos, info, errT)
+      ctx.slice.loc(getT, idxT, typT)(pos, info, errT)
+    }
+  }
+
+  /** An application of the "sconstruct[`typ`](...)" Viper function. */
+  private def construct(typ : vpr.Type, base : vpr.Exp, offset : vpr.Exp, len : vpr.Exp, cap : vpr.Exp)(ctx : Context)(pos : vpr.Position = vpr.NoPosition, info : vpr.Info = vpr.NoInfo, errT : vpr.ErrorTrafo = vpr.NoTrafos) : vpr.FuncApp =
+    constructGenerator(Vector(base, offset, len, cap), typ)(pos, info, errT)(ctx)
+
+  /** An application of the "sfullSliceFromArray[`typ`](...)" Viper function. */
+  private def fullSliceFromArray(typ : vpr.Type, base : vpr.Exp, i : vpr.Exp, j : vpr.Exp, k : vpr.Exp)(ctx : Context)(pos : vpr.Position = vpr.NoPosition, info : vpr.Info = vpr.NoInfo, errT : vpr.ErrorTrafo = vpr.NoTrafos) : vpr.FuncApp =
+    fullSliceFromArrayGenerator(Vector(base, i, j, k), typ)(pos, info, errT)(ctx)
+
+  /** An application of the "sfullSliceFromSlice[`typ`](...)" Viper function. */
+  private def fullSliceFromSlice(typ : vpr.Type, base : vpr.Exp, i : vpr.Exp, j : vpr.Exp, k : vpr.Exp)(ctx : Context)(pos : vpr.Position = vpr.NoPosition, info : vpr.Info = vpr.NoInfo, errT : vpr.ErrorTrafo = vpr.NoTrafos) : vpr.FuncApp =
+    fullSliceFromSliceGenerator(Vector(base, i, j, k), typ)(pos, info, errT)(ctx)
+
+  /** Gives the 'nil' slice of type `typ`. */
+  private def nilSlice(typ : vpr.Type)(ctx : Context)(pos : vpr.Position, info : vpr.Info, errT : vpr.ErrorTrafo) : vpr.Exp =
+    ctx.option.none(ctx.slice.typ(typ))(pos, info, errT)
```

Comment:
Can your encoding of the nil slice derive that a slice with permissions is not nil?
E.g. the following code should verify:
```
requies acc(&x[0])
func foo(x []int) {
  assert x != nil
}
```

Reply:
No. Slices are encoded as option types. And given any expression `e : Option[T]`, assuming something about `optGet(e)` does not imply that `e != optNone()`. For example:

```viper
method m(o : Option[Int])
{
  assert 0 < optGet(o) ==> o != optNone() // does not verify
}
```

That is the reason why your code snippet also will not verify.

(This is exactly what I tried to solve in my encoding of Option types which appeared unsound...)

Reply:
Hmm, but I could of course translate `acc(&s[i])` into something like `!isNone(s) && acc(...).val)`. What do you think of that?

Reply:
The whole nil business is quite complex. That is also why I intentionally did not pick option for shared arrays. 
Maybe you could reuse the nil from array. There I made it such that only a single representation of null exists, which is an array of size 1 with all nil entries. The size embeddings preserve nil, i.e. they do not enforce their respective size constraint as this would violate the uniqueness of nil (which is inconsistent together with injectivity). Furthermore I adapted that loc is only injective wihtin its bound (thus forcing that there is a single nil).

If you want to keep the options, then maybe an axiom along the lines of the following example might help:
```
Slice[T] {
  ....
  axiom {
    forall s: Option[Slice[T]], i: Int ::     loc(sarray(get(s)), i) != (nil(); T) ==> s != none()
  }
}
```
This would require axiomatizing all our nil representations, which is a bit slow, but possible, e.g.
```
domain Nil[T] {
  function nil(): T

  axiom { (nil(): Ref) == null }
  axiom { forall i: Int :: {loc((nil(): Array[T]), i)} loc((nil(): Array[T]), i) == (nil(): T) }
  axiom { (nil(): Emb17[T]) == box17((nil(): Array[T])) && unbox17((nil(): Emb17[T])) == (nil(): Array[T]) }
  axiom { (nil(): Emb42[T]) == box42((nil(): Array[T])) && unbox42((nil(): Emb42[T])) == (nil(): Array[T]) }
  axiom { (nil(): Emb77[T]) == box77((nil(): Array[T])) && unbox77((nil(): Emb77[T])) == (nil(): Array[T]) }
}
```

Reply:
Yes, having such a different representation for `nil` slices could very well work, thanks!

I considered alternatives, but picked option types since they allow straightforward encoding of e.g. `len(s)` and `cap(s)` in case `s = nil`. But perhaps option types aren't the best choice after all...

I think the `nil` slice can alternatively be represented as a slice that has the `nil` array as its underlying array, and appropriate offset/len/cap values. For example `sconstruct(nilArray(), 0, 1, 1)`. Such a slice would be unique.

Reply:
I think length and cap also have to be 0. I wrote something about this further down.

Reply:
Yes, I agree

Reply:
I should also point out the following:
Go distinguishes between a slice of nil and a slice of an array of size 0, i.e. `var x []int = nil; var y [0]int; assert x[:] != y[:]`

Reply:
Hmm, yes. Go wouldn't accept that, since slices are in principle incomparable. But on the (Viper) encoding level these should indeed be different.

Reply:
Could we quickly have a zoom meeting?